File: android/src/main/java/com/rnwavtoflac/RnWavToFlacModule.java
package com.rnwavtoflac;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.module.annotations.ReactModule;

import android.util.Log;

@ReactModule(name = RnWavToFlacModule.NAME)
public class RnWavToFlacModule extends ReactContextBaseJavaModule {
  
    public static final String NAME = "RnWavToFlac";

    public RnWavToFlacModule(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @Override
    @NonNull
    public String getName() {
        return NAME;
    }

    static {
        final String nativeLibName = "WavToFlac";  // Defined in android/CMakeList.txt; keep in sync
        try {
            // Load 'native-lib' library on application startup.
            System.loadLibrary(nativeLibName);
            Log.e(NAME, "Loaded native library: " + nativeLibName);
        } 
        catch (Exception e) {
            Log.e(NAME, "Unable to load native library: " + nativeLibName);
        }
    }

    // Native method signatures underpinning react methods; see cpp-adapter.cpp for definitions

    public static native double nativeMultiply(double a, double b);
    public static native int nativeWavToFlac(String inputWavFilePath, String outputFlacFilePath);
    
    // React methods

    @ReactMethod
    public void multiply(double a, double b, Promise promise) {
        promise.resolve(nativeMultiply(a, b));
    }

    @ReactMethod
    public void wavToFlac(String inputWavFilePath, String outputFlacFilePath, Promise promise) {
        promise.resolve(nativeWavToFlac(inputWavFilePath, outputFlacFilePath));
    }
}

File: android/src/main/java/com/rnwavtoflac/RnWavToFlacModule.kt
package com.rnwavtoflac

import android.util.Log

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.Promise

// For notes on calling C++ from Kotlin, see: 
// https://medium.com/@mregnauld/how-to-run-c-code-from-a-native-android-or-ios-app-265cad775ed5

class RnWavToFlacModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

  companion object {
    const val NAME = "RnWavToFlac"
    init {
      System.loadLibrary("rn-wav-to-flac")
    }
    // NDK functions, here declared as static:
    @JvmStatic private external fun nativeMultiply(a:Double, b:Double): Double
    @JvmStatic private external fun nativeWavToFlac(inputWavFilePath:String, outputFlacFilePath:String): Int
    @JvmStatic private external fun nativeGetWavMetaData(inputWavFilePath:String): String
  }

  override fun getName(): String {
    return NAME
  }



  // React methods

  // Example method
  // See https://reactnative.dev/docs/native-modules-android
  @ReactMethod fun multiply(a: Double, b: Double, promise: Promise) {
    promise.resolve(nativeMultiply(a, b))
  }

  @ReactMethod fun getWavMetaData(inputWavFilePath:String, promise:Promise) {
    promise.resolve(nativeGetWavMetaData(inputWavFilePath))
  }

  @ReactMethod fun wavToFlac(inputWavFilePath:String, outputFlacFilePath:String, promise:Promise) {
    promise.resolve(nativeWavToFlac(inputWavFilePath, outputFlacFilePath))
  }
}
